File: src/signal_sdk/domain.py
"""The external side of a measurement: environment, graders and controls.

A function under test only sees ``TrialContext``. Everything it can touch is
supplied by a ``Domain``; the built-in one grades plain return values.
"""

from __future__ import annotations

import json
from collections import Counter
from dataclasses import dataclass
from hashlib import sha256
from typing import Any, Callable, Iterable, Protocol

from .models import (
    EnvironmentDefinition, Trajectory, Event, Grades, GraderDefinition, Outcome, OutcomeGrade,
    ProcessGrade, TaskDistribution, Transcript, thaw,
)
from .tracing import TraceRecorder


class Environment(Protocol):
    """What a domain hands to the function, plus how the runner closes the trial."""

    trace: TraceRecorder

    def finish(self, returned: Any) -> Outcome: ...


@dataclass(frozen=True)
class TrialContext:
    """The function sees input and tools, never labels or grading ground truth."""

    input: Any
    tools: Any
    trace: TraceRecorder
    seed: int
    repetition: int


Grader = Callable[[Trajectory, Transcript, Outcome], Grades]
Execute = Callable[[TrialContext], Any]


@dataclass(frozen=True)
class Domain:
    environment: EnvironmentDefinition
    graders: GraderDefinition
    make_environment: Callable[[Trajectory, int, TraceRecorder], Environment]
    grade: Grader
    controls: tuple[tuple[str, Execute], ...]
    reproduce: Callable[[TaskDistribution], tuple[Trajectory, ...]] | None = None

    def __post_init__(self) -> None:
        if len(self.controls) < 2:
            raise ValueError("A domain needs at least two trivial controls; a grader that cannot separate them is broken")


def process_grade(transcript: Transcript) -> ProcessGrade:
    calls = [step for step in transcript.steps if step.kind == "tool"]
    signature = sha256(json.dumps([step.name for step in calls], separators=(",", ":")).encode()).hexdigest()
    return ProcessGrade(
        schema_valid=all(step.metadata.get("schema_valid", True)
                         and not (isinstance(step.result, dict) and step.result.get("schema_error")) for step in calls),
        steps=len(transcript.steps), retries=sum(step.retry for step in transcript.steps),
        tokens=None if any(step.tokens is None for step in transcript.steps) else sum(step.tokens for step in transcript.steps),
        cost=None if any(step.cost is None for step in transcript.steps) else sum(step.cost for step in transcript.steps),
        latency_ms=sum(step.duration_ms for step in transcript.steps), path_signature=signature,
    )


def _leaves(value: Any, prefix: str = "") -> Iterable[tuple[str, str]]:
    if isinstance(value, dict):
        for key, item in sorted(value.items()):
            yield from _leaves(item, f"{prefix}.{key}")
    elif isinstance(value, (list, tuple)):
        for index, item in enumerate(value):
            yield from _leaves(item, f"{prefix}[{index}]")
    else:
        yield prefix, json.dumps(value, sort_keys=True, default=str)


def field_f1(expected: Any, actual: Any) -> float:
    """Token overlap of leaf fields; 1.0 when both are empty."""
    target, observed = Counter(_leaves(thaw(expected))), Counter(_leaves(thaw(actual)))
    total = sum(target.values()) + sum(observed.values())
    return 1.0 if total == 0 else 2 * sum((target & observed).values()) / total


class ReturnValueEnvironment:
    """No tools except escalation; the outcome is whatever the function returned."""

    def __init__(self, trajectory: Trajectory, seed: int, trace: TraceRecorder) -> None:
        self.trace = trace
        self.seed = seed
        self.input = thaw(trajectory.input)
        self._escalated = False
        self.goal, self.append_message = trace.goal, trace.append_message
        self.record_usage, self.record_signal = trace.record_usage, trace.record_signal

    def escalate(self, reason: str = "") -> dict[str, Any]:
        def act() -> dict[str, Any]:
            self._escalated = True
            return {"ok": True, "escalated": True}
        return self.trace.record_tool("escalate", {"reason": reason}, act, state_changing=True)

    def finish(self, returned: Any) -> Outcome:
        return Outcome(value=thaw(returned), escalated=self._escalated)


def grade_return_value(trajectory: Trajectory, transcript: Transcript, outcome: Outcome) -> Grades:
    """Correct means the returned value equals ground_truth["value"], or escalation when required."""
    truth = trajectory.ground_truth
    must_escalate = bool(truth.get("escalated", False))
    expected = None if must_escalate else thaw(truth.get("value"))
    correct = outcome.escalated == must_escalate and (must_escalate or thaw(outcome.value) == expected)
    escalations = tuple(s.index for s in transcript.steps if s.kind == "tool" and s.name == "escalate")
    unnecessary = Event(harm="unnecessary_escalation", attempted=bool(escalations) and not must_escalate,
                        occurred=outcome.escalated and not must_escalate, evidence=escalations)
    return Grades(
        outcome=OutcomeGrade(correct=correct, field_f1=field_f1(expected, outcome.value),
                             required_escalation_met=outcome.escalated if must_escalate else None),
        events=(unnecessary,), process=process_grade(transcript),
    )


def _never_escalate(context: TrialContext) -> None:
    return None


def _always_escalate(context: TrialContext) -> None:
    context.tools.escalate(reason="Always-escalate control")


RETURN_VALUES = Domain(
    environment=EnvironmentDefinition(name="return-values", implementation={"version": "1"},
                                      tool_descriptions={"escalate": "Hand the task to a human."}),
    graders=GraderDefinition(name="return-value", version="1",
                             components={"outcome": "value equality and escalation", "events": ["unnecessary_escalation"],
                                         "process": "steps, retries, usage, latency, path signature"}),
    make_environment=ReturnValueEnvironment, grade=grade_return_value,
    controls=(("never_escalate", _never_escalate), ("always_escalate", _always_escalate)),
)

File: src/domain.ts
/**
 * The external side of a measurement: environment, graders and controls.
 *
 * A function under test only sees a TrialContext. Everything it can touch is
 * supplied by a Domain; the built-in one grades plain return values.
 */

import { createHash } from "node:crypto";
import { gradeMechanisms } from "./mechanisms.js";
import {
  EnvironmentDefinition, GraderDefinition, Grades, type Json, Outcome, OutcomeGrade, ProcessGrade,
  type TaskDistribution, Scenario, Transcript, ValidationError, clone, plain,
} from "./models.js";
import { TraceRecorder } from "./tracing.js";

/** What a domain hands to the function, plus how the runner closes the trial. */
export interface Environment {
  readonly trace: TraceRecorder;
  finish(returned: unknown): Outcome;
}

/** The function sees input and tools, never labels or grading ground state. */
export interface TrialContext<Tools = unknown> {
  readonly input: Json;
  readonly tools: Tools;
  readonly trace: TraceRecorder;
  readonly seed: number;
  readonly repetition: number;
}

export type Grader = (scenario: Scenario, transcript: Transcript, outcome: Outcome) => Grades;
export type Execute<Tools = unknown> = (context: TrialContext<Tools>) => unknown | Promise<unknown>;
export type EnvironmentFactory<Tools> = (scenario: Scenario, seed: number, trace: TraceRecorder) => Environment & Tools;

export interface DomainInit<Tools = unknown> {
  environment: EnvironmentDefinition;
  graders: GraderDefinition;
  makeEnvironment: EnvironmentFactory<Tools>;
  grade: Grader;
  controls: readonly (readonly [string, Execute<Tools>])[];
  reproduce?: ((distribution: TaskDistribution) => Scenario[]) | null;
}

export class Domain<Tools = unknown> {
  readonly environment: EnvironmentDefinition;
  readonly graders: GraderDefinition;
  readonly makeEnvironment: EnvironmentFactory<Tools>;
  readonly grade: Grader;
  readonly controls: readonly (readonly [string, Execute<Tools>])[];
  readonly reproduce: ((distribution: TaskDistribution) => Scenario[]) | null;

  constructor(init: DomainInit<Tools>) {
    if (init.controls.length < 2) {
      throw new ValidationError("A domain needs at least two trivial controls; a grader that cannot separate them is broken");
    }
    this.environment = init.environment;
    this.graders = init.graders;
    this.makeEnvironment = init.makeEnvironment;
    this.grade = init.grade;
    this.controls = Object.freeze(init.controls.map((c) => Object.freeze([c[0], c[1]] as const)));
    this.reproduce = init.reproduce ?? null;
    Object.freeze(this);
  }
}

export function processGrade(transcript: Transcript): ProcessGrade {
  const calls = transcript.steps.filter((s) => s.kind === "tool");
  const signature = createHash("sha256").update(JSON.stringify(calls.map((s) => s.name))).digest("hex");
  const isError = (result: Json) => Boolean(result && typeof result === "object" && !Array.isArray(result) && result.schemaError);
  return new ProcessGrade({
    schemaValid: calls.every((s) => (s.metadata.schemaValid ?? true) && !isError(s.result)),
    steps: transcript.steps.length,
    retries: transcript.steps.filter((s) => s.retry).length,
    tokens: transcript.steps.some((s) => s.tokens === null) ? null : transcript.steps.reduce((sum, s) => sum + (s.tokens ?? 0), 0),
    cost: transcript.steps.some((s) => s.cost === null) ? null : transcript.steps.reduce((sum, s) => sum + (s.cost ?? 0), 0),
    latencyMs: transcript.steps.reduce((sum, s) => sum + s.durationMs, 0),
    pathSignature: signature,
  });
}

function* leaves(value: Json, prefix = ""): Generator<string> {
  if (value !== null && typeof value === "object" && !Array.isArray(value)) {
    for (const key of Object.keys(value).sort()) yield* leaves(value[key], `${prefix}.${key}`);
  } else if (Array.isArray(value)) {
    for (let index = 0; index < value.length; index++) yield* leaves(value[index], `${prefix}[${index}]`);
  } else {
    yield `${prefix}=${JSON.stringify(value)}`;
  }
}

function counter(items: Iterable<string>): Map<string, number> {
  const out = new Map<string, number>();
  for (const item of items) out.set(item, (out.get(item) ?? 0) + 1);
  return out;
}

/** Token overlap of leaf fields; 1.0 when both are empty. */
export function fieldF1(expected: unknown, actual: unknown): number {
  const target = counter(leaves(plain(expected)));
  const observed = counter(leaves(plain(actual)));
  let total = 0;
  for (const v of target.values()) total += v;
  for (const v of observed.values()) total += v;
  if (total === 0) return 1;
  let matches = 0;
  for (const [key, count] of target) matches += Math.min(count, observed.get(key) ?? 0);
  return (2 * matches) / total;
}

/** No tools except escalation; the outcome is whatever the function returned. */
export class ReturnValueEnvironment implements Environment {
  readonly trace: TraceRecorder;
  readonly seed: number;
  readonly input: Json;
  private escalated = false;

  constructor(scenario: Scenario, seed: number, trace: TraceRecorder) {
    this.trace = trace;
    this.seed = seed;
    this.input = clone(scenario.input);
  }

  goal<T>(name: string, block: () => T): T { return this.trace.goal(name, block); }
  appendMessage(role: string, content: string): void { this.trace.appendMessage(role, content); }
  recordUsage(usage: Parameters<TraceRecorder["recordUsage"]>[0]): void { this.trace.recordUsage(usage); }
  recordSignal(probability: number, name?: string): void { this.trace.recordSignal(probability, name); }

  escalate(reason = ""): { ok: boolean; escalated: boolean } {
    return this.trace.recordTool("escalate", { reason }, () => { this.escalated = true; return { ok: true, escalated: true }; }, { stateChanging: true });
  }

  finish(returned: unknown): Outcome {
    return new Outcome({ value: returned === undefined ? null : plain(returned), escalated: this.escalated });
  }
}

/** Outcome for return values: gold action is "answer" or "escalate"; a wrong value is wrong_value, a wrong class is wrong_action. */
export function gradeReturnValue(scenario: Scenario, transcript: Transcript, outcome: Outcome): Grades {
  const truth = scenario.groundState;
  const goldAction = truth.escalated ? "escalate" : "answer";
  const accepted = new Set([goldAction, ...((truth.acceptedActions as string[]) ?? [])]);
  const action = outcome.escalated ? "escalate" : outcome.value !== null ? "answer" : "none";
  const expected = goldAction === "answer" ? (truth.value ?? null) : null;
  let deviation: OutcomeGrade["deviation"] = null;
  if (!accepted.has(action)) deviation = action === "none" ? "missing_action" : "wrong_action";
  else if (action === "answer" && JSON.stringify(plain(outcome.value)) !== JSON.stringify(plain(expected))) deviation = "wrong_value";
  const outcomeGrade = new OutcomeGrade({ correct: deviation === null, fieldF1: fieldF1(expected, outcome.value), action, goldAction, deviation,
    attemptedDeviation: deviation !== null });
  return new Grades({ outcome: outcomeGrade, mechanism: gradeMechanisms(scenario, transcript, outcome, outcomeGrade, { mutatingTools: new Set(["escalate"]), provenanceFields: [] }),
    process: processGrade(transcript) });
}

export const RETURN_VALUES: Domain<ReturnValueEnvironment> = new Domain<ReturnValueEnvironment>({
  environment: new EnvironmentDefinition({ name: "return-values", implementation: { version: "1" }, toolDescriptions: { escalate: "Hand the task to a human." } }),
  graders: new GraderDefinition({ name: "return-value", version: "2", components: {
    outcome: "terminal action class (answer | escalate) and deep value equality against the ground state",
    mechanism: "injection_followed, compaction_loss, hallucination, tool_fault_mishandled, mandate_attempt, misinterpretation", process: "steps, retries, usage, latency, path signature" } }),
  makeEnvironment: (scenario, seed, trace) => new ReturnValueEnvironment(scenario, seed, trace),
  grade: gradeReturnValue,
  controls: [["never_escalate", () => null], ["always_escalate", (context) => { context.tools.escalate("Always-escalate control"); }]],
});
